ENG-13196: Improve LTT block layout

wolffcm opened the pull request:
Here's what I did:
- Changed LargeTempTableBlock to put fixed-size tuple data in the front of its buffer and variable-size data in the back.
- Added comments to TupleBlock.h
- Removed the reference to TBPtr in TableIterator (because LTT blocks do not have a TBPtr)
- Changes to allow creation of StringRefs in LTT blocks, and not just voltdb::Pool.  I did this by creating function templates for functions that previously would only take a Pool.
- I needed to change the signatures of some common methods in NValue and TableTuple, so I updated the callers.
- Improved EE unit test for LargeTempTable, including an accounting of how memory is used

yzhang1991 commented:
Reviewed 24 of 25 files at r1.
Review status: 24 of 25 files reviewed at latest revision, 27 unresolved discussions.

---

*[src/ee/common/LargeTempTableBlockCache.cpp, line 69 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwCLoulH_1wtA6U9sdr:-KwCLoulH_1wtA6U9sds:b-3xslm8) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/LargeTempTableBlockCache.cpp#L69)):*
> ```cpp
>     if (! (*listIt)->isResident()) {
>         // If needed store a resident block to make space
>         while (m_totalAllocatedBytes + LargeTempTableBlock::BLOCK_SIZE_IN_BYTES > m_maxCacheSizeInBytes) {
> ```

Looks like this logic can be refactored as a reusable function.

---

*[src/ee/common/LargeTempTableBlockCache.cpp, line 73 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHEPrVf4SxvPiiqNha:-KwHEPrVf4SxvPiiqNhb:b-m7ck4s) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/LargeTempTableBlockCache.cpp#L73)):*
> ```cpp
>         }
> 
>         Topend* topend = ExecutorContext::getExecutorContext()->getTopend();
> ```

What do you think if we make this Topend* a member in the block cache? It is needed everywhere. Do you think it will be better if we do not have to call the function to get the Topend* every time when we need it?

---

*[src/ee/common/LargeTempTableBlockCache.cpp, line 141 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwCMiWk7B1as1usSsv7:-KwCMiWk7B1as1usSsv8:b-q9f9jy) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/LargeTempTableBlockCache.cpp#L141)):*
> ```cpp
>         BOOST_FOREACH (auto& block, m_blockList) {
>             if (block->isPinned()) {
>                 block->unpin();
> ```

Why the block has to be unpinned first here? You did not unpin the block in `releaseBlock()`.

---

*[src/ee/common/LargeTempTableBlockCache.cpp, line 154 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHFAFuD9O7EPFhSqJm:-KwHFAFuD9O7EPFhSqJn:b57q3nv) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/LargeTempTableBlockCache.cpp#L154)):*
> ```cpp
>         }
>         m_blockList.clear();
>         assert (m_totalAllocatedBytes == 0);
> ```

Move this to the outside of the if block

---

*[src/ee/common/LargeTempTableBlockCache.cpp, line 157 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHFE68dSK2kRoM7Bhi:-KwHFE68dSK2kRoM7Bhj:b-d8rdv8) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/LargeTempTableBlockCache.cpp#L157)):*
> ```cpp
>     }
>     else {
>         assert (m_totalAllocatedBytes == 0);
> ```

Remove the else block

---

*[src/ee/common/LargeTempTableBlockCache.cpp, line 164 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwCNvGOLs1gEhTNfZcO:-KwCNvGOLs1gEhTNfZcP:b-acplpm) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/LargeTempTableBlockCache.cpp#L164)):*
> ```cpp
> 
>     if (m_blockList.empty()) {
>         throwDynamicSQLException("LTT block cache requested a block be stored but there are no blocks");
> ```

`assert (m_totalAllocatedBytes == 0);`?

---

*[src/ee/common/LargeTempTableBlockCache.cpp, line 164 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHFLXgId_-6RfdJu50:-KwHFLXgId_-6RfdJu51:b-3b73zy) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/LargeTempTableBlockCache.cpp#L164)):*
> ```cpp
> 
>     if (m_blockList.empty()) {
>         throwDynamicSQLException("LTT block cache requested a block be stored but there are no blocks");
> ```

Throughout your code here you always use "DynamicSQLException". However, to me, it does not sound like a "Dynamic SQL Exception". It is an exception at the storage level.
Do you think there is a better exception class for this?

---

*[src/ee/common/NValue.hpp, line 301 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHGdO3mwQJSHfCOpvD:-KwHGdO4hOmgKq1spQVL:bew8fkd) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/NValue.hpp#L301)):*
> ```cpp
> 
>     /** This method is similar to the one above, but accepts no pool
>         argument.  If allocation is requested, objects will be copied
> ```

If allocation is requested (allocateObjects = true)

---

*[src/ee/common/NValue.hpp, line 750 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHH562CxL-odRcFTzf:-KwHH562CxL-odRcFTzg:b-n2yqyu) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/NValue.hpp#L750)):*
> ```cpp
>     }
> 
>     /** Return the amount of memory need to store this outlined value
> ```

needed

---

*[src/ee/common/NValue.hpp, line 763 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHH6TQiIMaHOArc0Ch:-KwHH6TQiIMaHOArc0Ci:b-n2yqyu) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/NValue.hpp#L763)):*
> ```cpp
>     }
> 
>     /** Return the amount of memory need to store this outlined value
> ```

needed

---

*[src/ee/common/StringRef.h, line 43 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHIJZdbio7g11ZDB9y:-KwHIJZdbio7g11ZDB9z:b2il82d) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/StringRef.h#L43)):*
> ```objc
>     int32_t getAllocatedSizeInPersistentStorage() const;
> 
>     /// The method is just like getAllocatedSize() but it returns the
> ```

just like `getAllocatedSizeInPersistentStorage()` but...

---

*[src/ee/common/tabletuple.h, line 191 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHJ5B0WKptbc1l_2GY:-KwHJ5B0WKptbc1l_2GZ:btswbrc) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/tabletuple.h#L191)):*
> ```objc
>     }
> 
>     /** Return the amount of memory needed to store the non-inlined
> ```

I suggest unifying the names throughout this pull request.
In this pull request, I can see you used "**UninIined**", "**NonInlined**", and "**Outlined**". For example, "getUninlinedObjectColumnCount", "getNonInlinedMemorySizeForPersistentTable"

To me "Outlined" or "NonInlined" sounds more legit.
Either one is fine, but not mixing them all together, here and there.

---

*[src/ee/common/tabletuple.h, line 219 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHKxLnlIbJdc1KM1qk:-KwHKxLnlIbJdc1KM1ql:b-qteb26) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/tabletuple.h#L219)):*
> ```objc
>         size_t bytes = 0;
>         uint16_t outlinedColCount = m_schema->getUninlinedObjectColumnCount();
>         for (uint16_t ii = 0; ii < outlinedColCount; ii++) {
> ```

Why `ii`?

---

*[src/ee/common/tabletuple.h, line 308 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHLBFeX5xl_LaxzK2D:-KwHLBFeX5xl_LaxzK2E:b-xca14t) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/tabletuple.h#L308)):*
> ```objc
>      * to provide NULL for stringPool in which case the strings will
>      * be allocated in persistent, relocatable storage.
>      * The POOL argument may either be an instance to a Pool object or
> ```

a Pool instance
or 
a pointer to a Pool object

---

*[src/ee/storage/LargeTempTable.cpp, line 37 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHNBLW4UUSMaFrz4rc:-KwHNBLW4UUSMaFrz4rd:b4lorrv) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTable.cpp#L37)):*
> ```cpp
> 
>     if (m_blockForWriting != NULL) {
>         lttBlockCache->unpinBlock(m_blockForWriting->id());
> ```

Why we do not just use
```
m_blockForWriting->unpin();
```
It saves the search time, isn't it?

---

*[src/ee/storage/LargeTempTable.cpp, line 49 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwH6ZWsg8H3WFlxZxQg:-KwH6ZWsg8H3WFlxZxQh:b5wjqu9) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTable.cpp#L49)):*
> ```cpp
>     assert(! m_insertsFinished);
> 
>     if (m_blockForWriting == NULL) {
> ```

If we start with `m_blockForWriting == NULL` and both insert tuple attempts have `success == false`, then you allocate two empty blocks and the first one has nothing in it and is no longer the `blockForWriting`.
That block is then wasted, right?

---

*[src/ee/storage/LargeTempTableBlock.cpp, line 34 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHPIKht267cVHQRwxT:-KwHPIKht267cVHQRwxU:brs8t1c) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTableBlock.cpp#L34)):*
> ```cpp
> 
> bool LargeTempTableBlock::insertTuple(const TableTuple& source) {
>     assert (m_tupleInsertionPoint < m_outlinedInsertionPoint);
> ```

<=
See below

---

*[src/ee/storage/LargeTempTableBlock.cpp, line 36 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHP1iWoBF5SaFhDBz6:-KwHP1iWoBF5SaFhDBz7:b5cq36z) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTableBlock.cpp#L36)):*
> ```cpp
>     assert (m_tupleInsertionPoint < m_outlinedInsertionPoint);
> 
>     size_t outlinedMemorySize = source.getNonInlinedMemorySizeForTempTable();
> ```

See the naming discrepancy?

---

*[src/ee/storage/LargeTempTableBlock.cpp, line 42 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHSTljH-_igeGzdldE:-KwHSTljH-_igeGzdldF:b-g5szb5) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTableBlock.cpp#L42)):*
> ```cpp
>     char* newOutlinedInsertionPoint = m_outlinedInsertionPoint - outlinedMemorySize;
> 
>     if (newTupleInsertionPoint > newOutlinedInsertionPoint) {
> ```

So after the tuple is inserted, it is possible that `m_tupleInsertionPoint == m_outlinedInsertionPoint`
That is why the assertion at the beginning of this function should be `<=` instead of `<`

---

*[src/ee/storage/LargeTempTableBlock.cpp, line 55 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHPotT7o_qu_JE0OxL:-KwHPotT7o_qu_JE0OxM:b-5jhrbw) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTableBlock.cpp#L55)):*
> ```cpp
>     ++m_activeTupleCount;
>     m_tupleInsertionPoint += target.tupleLength();
> ```

Add a comment here
m_outlinedInsertionPoint will be updated when `LargeTempTableBlock::allocate()` is called.

---

*[src/ee/storage/LargeTempTableBlock.cpp, line 66 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHQQ9wwVss21PwLq6O:-KwHQQ9wwVss21PwLq6P:b2nh6ex) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTableBlock.cpp#L66)):*
> ```cpp
> void* LargeTempTableBlock::allocate(std::size_t size) {
>     m_outlinedInsertionPoint -= size;
>     assert(m_tupleInsertionPoint < m_outlinedInsertionPoint);
> ```

<=?

---

*[src/ee/storage/LargeTempTableBlock.cpp, line 81 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHQrUrj9dQG3oNStIq:-KwHQrUrj9dQG3oNStIr:b-dlfzhn) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTableBlock.cpp#L81)):*
> ```cpp
> int64_t LargeTempTableBlock::getAllocatedTupleMemory() const {
>     if (isResident())
>         return m_tupleInsertionPoint - m_storage.get();
> ```

Generally, I would wrap the statements in a brace pair even if there's only one line.

---

*[src/ee/storage/LargeTempTableBlock.h, line 33 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHRCtl85lRBbHcKniU:-KwHRCtl85lRBbHcKniV:bpfe7g5) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTableBlock.h#L33)):*
> ```objc
>  *
>  * The lower-addressed memory of the buffer is used to store tuples of
>  * fixed size, similar to persistent table blocks.  The
> ```

fixed-size, similar-to-persistent-table

---

*[src/ee/storage/LargeTempTableBlock.h, line 37 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHRtv7Q3seeTN_ZxNi:-KwHRtv7Q3seeTN_ZxNj:b-dsykll) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTableBlock.h#L37)):*
> ```objc
>  * referenced in the tuples.
>  *
>  * As the block is inserted into tuple and outlined memory grow
> ```

As tuples are inserted into the block, tuple and outlined memory grow...

---

*[src/ee/storage/LargeTempTableBlock.h, line 50 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHTD9raF177SPLlQH-:-KwHTD9raF177SPLlQH0:buq1qxw) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTableBlock.h#L50)):*
> ```objc
> 
>     /** The size of all large temp table blocks */
>     static const size_t BLOCK_SIZE_IN_BYTES = 8 * 1024 * 1024; // 8 MB
> ```

I think we need more comments here.
The maximum size of a row is 2MB.  Both too large and too small block size can lead to space waste.
Small block size will waste space when tuples are large.
Large block size will waste space when the table and the tuples are small.

---

*[src/ee/storage/LargeTempTableBlock.h, line 60 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHUUnk7q2vEhwvbd18:-KwHUUnk7q2vEhwvbd19:b-1g9e1k) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTableBlock.h#L60)):*
> ```objc
>     }
> 
>     /** insert another tuple into this block */
> ```

Returns `true` if the insertion was successful

---

*[src/ee/storage/LargeTempTableBlock.h, line 144 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHV-ynjx_UZr4a_auH:-KwHV-ynjx_UZr4a_auI:b-m9xag4) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTableBlock.h#L144)):*
> ```objc
> 
>     /** Points to the address where the next outlined object will be inserted */
>     char* m_outlinedInsertionPoint;
> ```

Actually, m_outlinedInsertionPoint - [next outlined object size] will the position where the next outlined object will be inserted.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/voltdb/voltdb/4924)*
<!-- Sent from Reviewable.io -->

yzhang1991 commented:
Reviewed 1 of 25 files at r1.
Review status: all files reviewed at latest revision, 27 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/voltdb/voltdb/4924)*
<!-- Sent from Reviewable.io -->

yzhang1991 commented:
Reviewed 1 of 25 files at r1.
Review status: all files reviewed at latest revision, 27 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/voltdb/voltdb/4924)*
<!-- Sent from Reviewable.io -->

wolffcm commented:
Review status: 13 of 37 files reviewed at latest revision, 27 unresolved discussions.

---

*[src/ee/common/LargeTempTableBlockCache.cpp, line 69 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwCLoulH_1wtA6U9sdr:-KwI-aKSdileJyApEKtz:b-896fix) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/LargeTempTableBlockCache.cpp#L69)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

Looks like this logic can be refactored as a reusable function.
</blockquote></details>

Done.

---

*[src/ee/common/LargeTempTableBlockCache.cpp, line 73 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHEPrVf4SxvPiiqNha:-KwI0CuOdDFr3Z13ANNU:bjepk8x) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/LargeTempTableBlockCache.cpp#L73)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

What do you think if we make this Topend* a member in the block cache? It is needed everywhere. Do you think it will be better if we do not have to call the function to get the Topend* every time when we need it?
</blockquote></details>

Sure, why not.

---

*[src/ee/common/LargeTempTableBlockCache.cpp, line 141 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwCMiWk7B1as1usSsv7:-KwI5glKfs2pe50HmuxG:b-wf7bgw) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/LargeTempTableBlockCache.cpp#L141)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

Why the block has to be unpinned first here? You did not unpin the block in `releaseBlock()`.
</blockquote></details>

Good question.  I removed the statement and my tests still pass.

I feel like I should assert that the block is not pinned before releasing it (in both release methods)---client code has an obligation to unpin blocks when they are done with them, before they are released.  I've added an EE unit test for this too. 

---

*[src/ee/common/LargeTempTableBlockCache.cpp, line 154 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHFAFuD9O7EPFhSqJm:-KwI2O4XSHJhiESyOsgA:b-896fix) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/LargeTempTableBlockCache.cpp#L154)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

Move this to the outside of the if block
</blockquote></details>

Done.

---

*[src/ee/common/LargeTempTableBlockCache.cpp, line 157 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHFE68dSK2kRoM7Bhi:-KwI2OsE5gE7CbTgw597:b-896fix) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/LargeTempTableBlockCache.cpp#L157)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

Remove the else block
</blockquote></details>

Done.

---

*[src/ee/common/LargeTempTableBlockCache.cpp, line 164 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwCNvGOLs1gEhTNfZcO:-KwI3XqNGCqdgU_BpHQk:b-896fix) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/LargeTempTableBlockCache.cpp#L164)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

`assert (m_totalAllocatedBytes == 0);`?
</blockquote></details>

Done.

---

*[src/ee/common/LargeTempTableBlockCache.cpp, line 164 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHFLXgId_-6RfdJu50:-KwI3_IXSp3yvbQf5zSw:bfqeicl) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/LargeTempTableBlockCache.cpp#L164)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

Throughout your code here you always use "DynamicSQLException". However, to me, it does not sound like a "Dynamic SQL Exception". It is an exception at the storage level.
Do you think there is a better exception class for this?
</blockquote></details>

`SerializableEEException` seems to be better fit.  I've changed the code to use that instead.

---

*[src/ee/common/NValue.hpp, line 301 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHGdO3mwQJSHfCOpvD:-KwLMCBnJqDmzTlbP74v:b-896fix) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/NValue.hpp#L301)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

If allocation is requested (allocateObjects = true)
</blockquote></details>

Done.

---

*[src/ee/common/NValue.hpp, line 750 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHH562CxL-odRcFTzf:-KwLMBIhCJudCH8vZzBY:b-896fix) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/NValue.hpp#L750)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

needed
</blockquote></details>

Done.

---

*[src/ee/common/NValue.hpp, line 763 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHH6TQiIMaHOArc0Ch:-KwLMAZD5Ghhk3KAfpcj:b-896fix) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/NValue.hpp#L763)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

needed
</blockquote></details>

Done.

---

*[src/ee/common/StringRef.h, line 43 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHIJZdbio7g11ZDB9y:-KwLM9-J3snGpQENKQ4g:b-896fix) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/StringRef.h#L43)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

just like `getAllocatedSizeInPersistentStorage()` but...
</blockquote></details>

Done.

---

*[src/ee/common/tabletuple.h, line 191 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHJ5B0WKptbc1l_2GY:-KwLErdFOygl2HqBEc_2:b-1pcolo) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/tabletuple.h#L191)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

I suggest unifying the names throughout this pull request.
In this pull request, I can see you used "**UninIined**", "**NonInlined**", and "**Outlined**". For example, "getUninlinedObjectColumnCount", "getNonInlinedMemorySizeForPersistentTable"

To me "Outlined" or "NonInlined" sounds more legit.
Either one is fine, but not mixing them all together, here and there.

</blockquote></details>

Another one is "uninlineable".  I agree it would be good to have just one term to define this kind of data.  I am not a fan of "non-inlined", but that is probably most pervasive, so I will use that.
I was able to remove all references to "outlined" data.  There are still several references to "uninlineable" data in common method names with many callers, so I have left them for now.

---

*[src/ee/common/tabletuple.h, line 219 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHKxLnlIbJdc1KM1qk:-KwLLsvRUuBsieC1quDt:b-896fix) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/tabletuple.h#L219)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

Why `ii`?
</blockquote></details>

Done.

---

*[src/ee/common/tabletuple.h, line 308 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHLBFeX5xl_LaxzK2D:-KwLN0WPYKGEmzX_w_eZ:b-896fix) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/common/tabletuple.h#L308)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

a Pool instance
or 
a pointer to a Pool object
</blockquote></details>

Done.

---

*[src/ee/storage/LargeTempTable.cpp, line 37 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHNBLW4UUSMaFrz4rc:-KwLNaT266Wbd8A4PLuD:b-896fix) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTable.cpp#L37)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

Why we do not just use
```
m_blockForWriting->unpin();
```
It saves the search time, isn't it?

</blockquote></details>

Done.

---

*[src/ee/storage/LargeTempTable.cpp, line 49 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwH6ZWsg8H3WFlxZxQg:-KwLPv1ARqsU9uuYEA65:bpufxd) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTable.cpp#L49)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

If we start with `m_blockForWriting == NULL` and both insert tuple attempts have `success == false`, then you allocate two empty blocks and the first one has nothing in it and is no longer the `blockForWriting`.
That block is then wasted, right?
</blockquote></details>

Well, if `m_blockForWriting` is NULL then we will be inserting into an empty block.  If the insert still fails, we have a tuple that needs more that 8MB of space, and we will have to abort the query in any case.  I will clean up the logic here though.

---

*[src/ee/storage/LargeTempTableBlock.cpp, line 34 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHPIKht267cVHQRwxT:-KwLRHo_7wzhxmTzAVC2:b-896fix) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTableBlock.cpp#L34)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

<=
See below
</blockquote></details>

Done.

---

*[src/ee/storage/LargeTempTableBlock.cpp, line 36 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHP1iWoBF5SaFhDBz6:-KwLRISVq6IxA25EESgk:b-896fix) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTableBlock.cpp#L36)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

See the naming discrepancy?
</blockquote></details>

Done.

---

*[src/ee/storage/LargeTempTableBlock.cpp, line 42 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHSTljH-_igeGzdldE:-KwLRIuTqGq73OJs90ET:b-896fix) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTableBlock.cpp#L42)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

So after the tuple is inserted, it is possible that `m_tupleInsertionPoint == m_outlinedInsertionPoint`
That is why the assertion at the beginning of this function should be `<=` instead of `<`
</blockquote></details>

Done.

---

*[src/ee/storage/LargeTempTableBlock.cpp, line 55 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHPotT7o_qu_JE0OxL:-KwLS1Qg7rPPV-RJJDbU:b-896fix) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTableBlock.cpp#L55)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

Add a comment here
m_outlinedInsertionPoint will be updated when `LargeTempTableBlock::allocate()` is called.
</blockquote></details>

Done.

---

*[src/ee/storage/LargeTempTableBlock.cpp, line 66 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHQQ9wwVss21PwLq6O:-KwLS2PhvznumwA6MXog:b-896fix) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTableBlock.cpp#L66)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

<=?
</blockquote></details>

Done.

---

*[src/ee/storage/LargeTempTableBlock.cpp, line 81 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHQrUrj9dQG3oNStIq:-KwLS2yU1QtvURqaqCtn:b-896fix) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTableBlock.cpp#L81)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

Generally, I would wrap the statements in a brace pair even if there's only one line.
</blockquote></details>

Done.

---

*[src/ee/storage/LargeTempTableBlock.h, line 33 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHRCtl85lRBbHcKniU:-KwLSB6ORiWE3bQvrY5m:b-pp28uk) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTableBlock.h#L33)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

fixed-size, similar-to-persistent-table
</blockquote></details>

Hmm.  I don't think your proposed change conveys the meaning I intended.  I have updated the comment so it's clearer though.

---

*[src/ee/storage/LargeTempTableBlock.h, line 37 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHRtv7Q3seeTN_ZxNi:-KwLW72-eKVqtu74dBAY:b-896fix) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTableBlock.h#L37)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

As tuples are inserted into the block, tuple and outlined memory grow...
</blockquote></details>

Done.

---

*[src/ee/storage/LargeTempTableBlock.h, line 50 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHTD9raF177SPLlQH-:-KwLX9N7-HXoPDoqmgXA:b-896fix) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTableBlock.h#L50)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

I think we need more comments here.
The maximum size of a row is 2MB.  Both too large and too small block size can lead to space waste.
Small block size will waste space when tuples are large.
Large block size will waste space when the table and the tuples are small.
</blockquote></details>

Done.

---

*[src/ee/storage/LargeTempTableBlock.h, line 60 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHUUnk7q2vEhwvbd18:-KwLXZ2sbyQwXCtpraz8:b-896fix) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTableBlock.h#L60)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

Returns `true` if the insertion was successful
</blockquote></details>

Done.

---

*[src/ee/storage/LargeTempTableBlock.h, line 144 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwHV-ynjx_UZr4a_auH:-KwLYOdAyTUQzTCJhwA1:b-896fix) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTableBlock.h#L144)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

Actually, m_outlinedInsertionPoint - [next outlined object size] will the position where the next outlined object will be inserted.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/voltdb/voltdb/4924)*
<!-- Sent from Reviewable.io -->

yzhang1991 commented:
Reviewed 24 of 24 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/ee/common/LargeTempTableBlockCache.cpp, line 181 at r2](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwMgqEjAjzFVmixRKTv:-KwMgqEjAjzFVmixRKTw:b-27pqr7) ([raw file](https://github.com/voltdb/voltdb/blob/290c9d796fc26bc056815ea409d33081270e7b59/src/ee/common/LargeTempTableBlockCache.cpp#L181)):*
> ```cpp
>             assert (m_totalAllocatedBytes >= 0);
>             assert (! block->isResident());
>             assert (m_totalAllocatedBytes + LargeTempTableBlock::BLOCK_SIZE_IN_BYTES <= m_maxCacheSizeInBytes);
> ```

I don't see the necessity of this assertion.
m_totalAllocatedBytes was just subtracted by BLOCK_SIZE_IN_BYTES three lines above.
If we can ever fail this assertion, we must have been in trouble for a while.

---

*[src/ee/storage/LargeTempTable.cpp, line 78 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwWx9VRhVddKpLqSPsW:-KwWx9VRhVddKpLqSPsX:b-tovbzt) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTable.cpp#L78)):*
> ```cpp
>     int64_t id = m_blockIds.back();
>     if (lttBlockCache->blockIsPinned(id)) {
>         lttBlockCache->unpinBlock(id);
> ```

I am thinking... Can this just be unpinning the m_blockForWriting if it is not null?
If m_blockForWriting is not null when this method is called, should we set it to null here?

---

*[src/ee/storage/LargeTempTableBlock.cpp, line 89 at r2](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwWtK0cU_e_zW0wBvEN:-KwWtK0cU_e_zW0wBvEO:b-b36yoe) ([raw file](https://github.com/voltdb/voltdb/blob/290c9d796fc26bc056815ea409d33081270e7b59/src/ee/storage/LargeTempTableBlock.cpp#L89)):*
> ```cpp
> 
> int64_t LargeTempTableBlock::getAllocatedPoolMemory() const {
>     if (isResident())
> ```

Wrap it with a brace pair.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/voltdb/voltdb/4924)*
<!-- Sent from Reviewable.io -->

wolffcm commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/ee/common/LargeTempTableBlockCache.cpp, line 181 at r2](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwMgqEjAjzFVmixRKTv:-Kwa1pLMQ4gHU3RLsxvk:b-896fix) ([raw file](https://github.com/voltdb/voltdb/blob/290c9d796fc26bc056815ea409d33081270e7b59/src/ee/common/LargeTempTableBlockCache.cpp#L181)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

I don't see the necessity of this assertion.
m_totalAllocatedBytes was just subtracted by BLOCK_SIZE_IN_BYTES three lines above.
If we can ever fail this assertion, we must have been in trouble for a while.
</blockquote></details>

Done.

---

*[src/ee/storage/LargeTempTable.cpp, line 78 at r1](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwWx9VRhVddKpLqSPsW:-Kwaxww59Kt_A3srbg2F:bmis6u5) ([raw file](https://github.com/voltdb/voltdb/blob/c0c005332df619043ce6f60aa61ab0661801a261/src/ee/storage/LargeTempTable.cpp#L78)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

I am thinking... Can this just be unpinning the m_blockForWriting if it is not null?
If m_blockForWriting is not null when this method is called, should we set it to null here?
</blockquote></details>

Yes, I think you're right.  My next commit will have some simplification as it seems like m_insertsFinished is a bit of state that is not needed.  It can always be inferred from whether or not m_blockForWriting is NULL.

---

*[src/ee/storage/LargeTempTableBlock.cpp, line 89 at r2](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-KwWtK0cU_e_zW0wBvEN:-Kway89KkbS6eOa4Zz4T:b-896fix) ([raw file](https://github.com/voltdb/voltdb/blob/290c9d796fc26bc056815ea409d33081270e7b59/src/ee/storage/LargeTempTableBlock.cpp#L89)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

Wrap it with a brace pair.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/voltdb/voltdb/4924)*
<!-- Sent from Reviewable.io -->

yzhang1991 commented:
Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/ee/storage/LargeTempTable.cpp, line 43 at r3](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-Kwb0gqn_djdWpyYHLS8:-Kwb0gqn_djdWpyYHLS9:b-poodj9) ([raw file](https://github.com/voltdb/voltdb/blob/2516aecd598a7cddbf7ec4607ab3b249b0050da9/src/ee/storage/LargeTempTable.cpp#L43)):*
> ```cpp
>     // Try to get an empty block (this will invoke I/O via topend, and
>     // could throw for any number of reasons)
>     LargeTempTableBlock* newBlock = lttBlockCache->getEmptyBlock();
> ```

Although it is unlikely. if this throws, m_blockForWriting is still unpinned. Will it be a problem?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/voltdb/voltdb/4924)*
<!-- Sent from Reviewable.io -->

wolffcm commented:
*[src/ee/storage/LargeTempTable.cpp, line 43 at r3](https://reviewable.io:443/reviews/voltdb/voltdb/4924#-Kwb0gqn_djdWpyYHLS8:-Kwb32q_gLy-B451O9dC:b7bkwpg) ([raw file](https://github.com/voltdb/voltdb/blob/2516aecd598a7cddbf7ec4607ab3b249b0050da9/src/ee/storage/LargeTempTable.cpp#L43)):*
<details><summary><i>Previously, yzhang1991 (Yiqun Zhang) wrote…</i></summary><blockquote>

Although it is unlikely. if this throws, m_blockForWriting is still unpinned. Will it be a problem?
</blockquote></details>

No, I think it's fine.  if we throw here (not that unlikely---e.g, disk is full) then we will need to abort the query and there is no reason why the old block needs to be pinned.  I have relaxed finishInserts to reflect that the block may or may not be pinned.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/voltdb/voltdb/4924)*
<!-- Sent from Reviewable.io -->

yzhang1991 commented:
Reviewed 2 of 2 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/voltdb/voltdb/4924)*
<!-- Sent from Reviewable.io -->
